src/net_processing.cpp

```diff
@@ -253,9 +250,18 @@ class PeerManagerImpl final : public PeerManager
     void SetBestHeight(int height) override { m_best_height = height; };
     void Misbehaving(const NodeId pnode, const int howmuch, const std::string& message) override;
     void ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
-                        const std::chrono::microseconds time_received, const std::atomic<bool>& interruptMsgProc) override;
+                        const std::chrono::microseconds time_received, const std::atomic<bool>& interruptMsgProc) override LOCKS_EXCLUDED(m_mutex_messages);
 
 private:
+    /** Message processing is single-threaded
+     * Anything only accessed by message parsing can be guarded by
+     * this mutex, which is acquired by ProcessMessages and SendMessages
+     */
+    Mutex m_mutex_messages;
+
+    void _ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
```

Comment:
I think it'd be cleaner for the `m_mutex_messages` lock to be taken by ProcessMessage**s**, rather than ProcessMessage. That would remove the need to define an inner `_ProcessMessage` to be called when the mutex is already locked. It's also consistent with the comment on the `m_mutex_messages` member, which says "is acquired by ProcessMessages and SendMessages"

Reply:
It is taken by `ProcessMessages()` (and also by `SendMessages()`), however `ProcessMessage()` is also called from the test suite, so cannot assume the lock has been taken. That's why the public interface, `ProcessMessage()`, is implemented as a wrapper that takes the lock, then calls the internal function.

Reply:
Ah, sorry - misread.